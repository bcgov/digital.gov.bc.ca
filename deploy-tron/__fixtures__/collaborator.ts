export const adminUser = {
  permission: 'admin',
  user: {
    login: 'patricksimonian',
    id: 233,
    node_id: '13123',
    avatar_url: 'https://avatars1.githubusercontent.com/u/21046727?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/patricksimonian',
    html_url: 'https://github.com/patricksimonian',
    followers_url: 'https://api.github.com/users/patricksimonian/followers',
    following_url:
      'https://api.github.com/users/patricksimonian/following{/other_user}',
    gists_url: 'https://api.github.com/users/patricksimonian/gists{/gist_id}',
    starred_url:
      'https://api.github.com/users/patricksimonian/starred{/owner}{/repo}',
    subscriptions_url:
      'https://api.github.com/users/patricksimonian/subscriptions',
    organizations_url: 'https://api.github.com/users/patricksimonian/orgs',
    repos_url: 'https://api.github.com/users/patricksimonian/repos',
    events_url: 'https://api.github.com/users/patricksimonian/events{/privacy}',
    received_events_url:
      'https://api.github.com/users/patricksimonian/received_events',
    type: 'User',
    site_admin: false,
  },
};

export const readUser = {
  permission: 'read',
  user: {
    login: 'patricksimonian',
    id: 123,
    node_id: '3411',
    avatar_url: 'https://avatars1.githubusercontent.com/u/21046727?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/patricksimonian',
    html_url: 'https://github.com/patricksimonian',
    followers_url: 'https://api.github.com/users/patricksimonian/followers',
    following_url:
      'https://api.github.com/users/patricksimonian/following{/other_user}',
    gists_url: 'https://api.github.com/users/patricksimonian/gists{/gist_id}',
    starred_url:
      'https://api.github.com/users/patricksimonian/starred{/owner}{/repo}',
    subscriptions_url:
      'https://api.github.com/users/patricksimonian/subscriptions',
    organizations_url: 'https://api.github.com/users/patricksimonian/orgs',
    repos_url: 'https://api.github.com/users/patricksimonian/repos',
    events_url: 'https://api.github.com/users/patricksimonian/events{/privacy}',
    received_events_url:
      'https://api.github.com/users/patricksimonian/received_events',
    type: 'User',
    site_admin: false,
  },
};

export const triageUser = {
  permission: 'triage',
  user: {
    login: 'patricksimonian',
    id: 44,
    node_id: '3411',
    avatar_url: 'https://avatars1.githubusercontent.com/u/21046727?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/patricksimonian',
    html_url: 'https://github.com/patricksimonian',
    followers_url: 'https://api.github.com/users/patricksimonian/followers',
    following_url:
      'https://api.github.com/users/patricksimonian/following{/other_user}',
    gists_url: 'https://api.github.com/users/patricksimonian/gists{/gist_id}',
    starred_url:
      'https://api.github.com/users/patricksimonian/starred{/owner}{/repo}',
    subscriptions_url:
      'https://api.github.com/users/patricksimonian/subscriptions',
    organizations_url: 'https://api.github.com/users/patricksimonian/orgs',
    repos_url: 'https://api.github.com/users/patricksimonian/repos',
    events_url: 'https://api.github.com/users/patricksimonian/events{/privacy}',
    received_events_url:
      'https://api.github.com/users/patricksimonian/received_events',
    type: 'User',
    site_admin: false,
  },
};

export const writeUser = {
  permission: 'write',
  user: {
    login: 'patricksimonian',
    id: 21046727,
    node_id: '3411',
    avatar_url: 'https://avatars1.githubusercontent.com/u/21046727?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/patricksimonian',
    html_url: 'https://github.com/patricksimonian',
    followers_url: 'https://api.github.com/users/patricksimonian/followers',
    following_url:
      'https://api.github.com/users/patricksimonian/following{/other_user}',
    gists_url: 'https://api.github.com/users/patricksimonian/gists{/gist_id}',
    starred_url:
      'https://api.github.com/users/patricksimonian/starred{/owner}{/repo}',
    subscriptions_url:
      'https://api.github.com/users/patricksimonian/subscriptions',
    organizations_url: 'https://api.github.com/users/patricksimonian/orgs',
    repos_url: 'https://api.github.com/users/patricksimonian/repos',
    events_url: 'https://api.github.com/users/patricksimonian/events{/privacy}',
    received_events_url:
      'https://api.github.com/users/patricksimonian/received_events',
    type: 'User',
    site_admin: false,
  },
};

export const maintainUser = {
  permission: 'maintain',
  user: {
    login: 'patricksimonian',
    id: 4454,
    node_id: '3411',
    avatar_url: 'https://avatars1.githubusercontent.com/u/21046727?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/patricksimonian',
    html_url: 'https://github.com/patricksimonian',
    followers_url: 'https://api.github.com/users/patricksimonian/followers',
    following_url:
      'https://api.github.com/users/patricksimonian/following{/other_user}',
    gists_url: 'https://api.github.com/users/patricksimonian/gists{/gist_id}',
    starred_url:
      'https://api.github.com/users/patricksimonian/starred{/owner}{/repo}',
    subscriptions_url:
      'https://api.github.com/users/patricksimonian/subscriptions',
    organizations_url: 'https://api.github.com/users/patricksimonian/orgs',
    repos_url: 'https://api.github.com/users/patricksimonian/repos',
    events_url: 'https://api.github.com/users/patricksimonian/events{/privacy}',
    received_events_url:
      'https://api.github.com/users/patricksimonian/received_events',
    type: 'User',
    site_admin: false,
  },
};
