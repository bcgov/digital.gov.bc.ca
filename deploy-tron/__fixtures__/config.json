{
  "validGithubRoles": ["maintain", "admin", "write"],
  "botCommand": "/test",
  "microservices": ["postgres", "redis"],
  "environments": ["production", "development", "staging"],
  "environmentSynonyms": {
    "prod": "production",
    "dev": "development",
    "test": "staging"
  },
  "requiredEnvironments": {
    "development": [],
    "staging": ["development"],
    "production": ["staging", "development"]
  },
  "environmentsThatAllowConcurrentDeploys": ["development"]
}
