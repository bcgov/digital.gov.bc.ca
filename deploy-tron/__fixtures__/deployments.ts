export const deploymentStatusesPending = {
  "repository": {
    "deployments": {
      "edges": [
        {
          "node": {
            "latestStatus": "PENDING",
            "ref": {
              "name": "foo"
            }
          }
        }
      ]
    }
  }
};

export const deploymentStatusesPendingWithSameRef = {
  "repository": {
    "deployments": {
      "edges": [
        {
          "node": {
            "latestStatus": "PENDING",
            "ref": {
              "name": "foo"
            }
          }
        }
      ]
    }
  }
};

export const deploymentStatusesSuccess = {
  "repository": {
    "deployments": {
      "edges": [
        {
          "node": {
            "latestStatus": "SUCCESS",
            "ref": {
              "name": "foo"
            }
          }
        }
      ]
    }
  }
};

export const deploymentsForRef = [
  {
    "url": "https://api.github.com/repos/patricksimonian/strapi-demo/deployments/4",
    "id": 12,
    "node_id": "MDEwOkRlcGxveW1lbnQyNTk1MzE2ODQ=",
    "sha": "038dade51d78ccf77d396793410d54af438ca7ae",
    "ref": "patricksimonian-patch-1",
    "task": "deploy",
    "payload": "{\"microservice\":\"react\",\"pr\":11,\"user\":\"patricksimonian\"}",
    "original_environment": "development",
    "environment": "development",
    "description": "deploy request from ci-cd",
    "creator": {
      "login": "digital-gov-bc-ca-cicd-dev[bot]",
      "id": 343,
      "node_id": "MDM6Qm90Njk2MDIyOTY=",
      "avatar_url": "https://avatars0.githubusercontent.com/in/76688?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D",
      "html_url": "https://github.com/apps/digital-gov-bc-ca-cicd-dev",
      "followers_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/followers",
      "following_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/following{/other_user}",
      "gists_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/subscriptions",
      "organizations_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/orgs",
      "repos_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/repos",
      "events_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/events{/privacy}",
      "received_events_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/received_events",
      "type": "Bot",
      "site_admin": false
    },
    "created_at": "2020-08-17T21:47:07Z",
    "updated_at": "2020-08-17T21:47:07Z",
    "statuses_url": "https://api.github.com/repos/patricksimonian/strapi-demo/deployments/259531684/statuses",
    "repository_url": "https://api.github.com/repos/patricksimonian/strapi-demo",
    "performed_via_github_app": null
  },
  {
    "url": "https://api.github.com/repos/patricksimonian/strapi-demo/deployments/4",
    "id": 12,
    "node_id": "MDEwOkRlcGxveW1lbnQyNTk1MzE2ODQ=",
    "sha": "038dade51d78ccf77d396793410d54af438ca7ae",
    "ref": "patricksimonian-patch-1",
    "task": "deploy",
    "payload": "{\"microservice\":\"react\",\"pr\":11,\"user\":\"patricksimonian\"}",
    "original_environment": "development",
    "environment": "staging",
    "description": "deploy request from ci-cd",
    "creator": {
      "login": "digital-gov-bc-ca-cicd-dev[bot]",
      "id": 343,
      "node_id": "MDM6Qm90Njk2MDIyOTY=",
      "avatar_url": "https://avatars0.githubusercontent.com/in/76688?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D",
      "html_url": "https://github.com/apps/digital-gov-bc-ca-cicd-dev",
      "followers_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/followers",
      "following_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/following{/other_user}",
      "gists_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/subscriptions",
      "organizations_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/orgs",
      "repos_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/repos",
      "events_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/events{/privacy}",
      "received_events_url": "https://api.github.com/users/digital-gov-bc-ca-cicd-dev%5Bbot%5D/received_events",
      "type": "Bot",
      "site_admin": false
    },
    "created_at": "2020-08-17T21:47:07Z",
    "updated_at": "2020-08-17T21:47:07Z",
    "statuses_url": "https://api.github.com/repos/patricksimonian/strapi-demo/deployments/259531684/statuses",
    "repository_url": "https://api.github.com/repos/patricksimonian/strapi-demo",
    "performed_via_github_app": null
  }
];

export const deploymentStatusForDeploymentWhereAllSuccessful = [
  {
    "url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/123/statuses/23434",
    "id": 1312,
    "node_id": "MDE2OkRlcGxveW1lbnRTdGF0dXMzNjMzOTA1MzQ=",
    "state": "success",
    "creator": {
      "login": "patricksimonian",
      "id": 3434,
      "node_id": "MDQ6VXNlcjIxMDQ2NzI3",
      "avatar_url": "https://avatars1.githubusercontent.com/u/21046727?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/patricksimonian",
      "html_url": "https://github.com/patricksimonian",
      "followers_url": "https://api.github.com/users/patricksimonian/followers",
      "following_url": "https://api.github.com/users/patricksimonian/following{/other_user}",
      "gists_url": "https://api.github.com/users/patricksimonian/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/patricksimonian/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/patricksimonian/subscriptions",
      "organizations_url": "https://api.github.com/users/patricksimonian/orgs",
      "repos_url": "https://api.github.com/users/patricksimonian/repos",
      "events_url": "https://api.github.com/users/patricksimonian/events{/privacy}",
      "received_events_url": "https://api.github.com/users/patricksimonian/received_events",
      "type": "User",
      "site_admin": false
    },
    "description": "",
    "environment": "production",
    "target_url": "",
    "created_at": "2020-08-07T22:24:20Z",
    "updated_at": "2020-08-07T22:24:20Z",
    "deployment_url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/3434",
    "repository_url": "https://api.github.com/repos/bcgov/devhub-app-web",
    "performed_via_github_app": null
  },
  {
    "url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/3434/statuses/34434",
    "id": 34434,
    "node_id": "MDE2OkRlcGxveW1lbnRTdGF0dXMzNjMzOTA1MjA=",
    "state": "success",
    "creator": {
      "login": "patricksimonian",
      "id": 123,
      "node_id": "MDQ6VXNlcjIxMDQ2NzI3",
      "avatar_url": "https://avatars1.githubusercontent.com/u/21046727?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/patricksimonian",
      "html_url": "https://github.com/patricksimonian",
      "followers_url": "https://api.github.com/users/patricksimonian/followers",
      "following_url": "https://api.github.com/users/patricksimonian/following{/other_user}",
      "gists_url": "https://api.github.com/users/patricksimonian/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/patricksimonian/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/patricksimonian/subscriptions",
      "organizations_url": "https://api.github.com/users/patricksimonian/orgs",
      "repos_url": "https://api.github.com/users/patricksimonian/repos",
      "events_url": "https://api.github.com/users/patricksimonian/events{/privacy}",
      "received_events_url": "https://api.github.com/users/patricksimonian/received_events",
      "type": "User",
      "site_admin": false
    },
    "description": "",
    "environment": "staging",
    "target_url": "",
    "created_at": "2020-08-07T22:24:16Z",
    "updated_at": "2020-08-07T22:24:16Z",
    "deployment_url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/343",
    "repository_url": "https://api.github.com/repos/bcgov/devhub-app-web",
    "performed_via_github_app": null
  }
];

export const deploymentStatusForDeployment = [
  {
    "url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/123/statuses/23434",
    "id": 1312,
    "node_id": "MDE2OkRlcGxveW1lbnRTdGF0dXMzNjMzOTA1MzQ=",
    "state": "success",
    "creator": {
      "login": "patricksimonian",
      "id": 3434,
      "node_id": "MDQ6VXNlcjIxMDQ2NzI3",
      "avatar_url": "https://avatars1.githubusercontent.com/u/21046727?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/patricksimonian",
      "html_url": "https://github.com/patricksimonian",
      "followers_url": "https://api.github.com/users/patricksimonian/followers",
      "following_url": "https://api.github.com/users/patricksimonian/following{/other_user}",
      "gists_url": "https://api.github.com/users/patricksimonian/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/patricksimonian/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/patricksimonian/subscriptions",
      "organizations_url": "https://api.github.com/users/patricksimonian/orgs",
      "repos_url": "https://api.github.com/users/patricksimonian/repos",
      "events_url": "https://api.github.com/users/patricksimonian/events{/privacy}",
      "received_events_url": "https://api.github.com/users/patricksimonian/received_events",
      "type": "User",
      "site_admin": false
    },
    "description": "",
    "environment": "production",
    "target_url": "",
    "created_at": "2020-08-07T22:24:20Z",
    "updated_at": "2020-08-07T22:24:20Z",
    "deployment_url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/3434",
    "repository_url": "https://api.github.com/repos/bcgov/devhub-app-web",
    "performed_via_github_app": null
  },
  {
    "url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/3434/statuses/34434",
    "id": 34434,
    "node_id": "MDE2OkRlcGxveW1lbnRTdGF0dXMzNjMzOTA1MjA=",
    "state": "pending",
    "creator": {
      "login": "patricksimonian",
      "id": 123,
      "node_id": "MDQ6VXNlcjIxMDQ2NzI3",
      "avatar_url": "https://avatars1.githubusercontent.com/u/21046727?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/patricksimonian",
      "html_url": "https://github.com/patricksimonian",
      "followers_url": "https://api.github.com/users/patricksimonian/followers",
      "following_url": "https://api.github.com/users/patricksimonian/following{/other_user}",
      "gists_url": "https://api.github.com/users/patricksimonian/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/patricksimonian/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/patricksimonian/subscriptions",
      "organizations_url": "https://api.github.com/users/patricksimonian/orgs",
      "repos_url": "https://api.github.com/users/patricksimonian/repos",
      "events_url": "https://api.github.com/users/patricksimonian/events{/privacy}",
      "received_events_url": "https://api.github.com/users/patricksimonian/received_events",
      "type": "User",
      "site_admin": false
    },
    "description": "",
    "environment": "development",
    "target_url": "",
    "created_at": "2020-08-07T22:24:16Z",
    "updated_at": "2020-08-07T22:24:16Z",
    "deployment_url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/343",
    "repository_url": "https://api.github.com/repos/bcgov/devhub-app-web",
    "performed_via_github_app": null
  }
];
