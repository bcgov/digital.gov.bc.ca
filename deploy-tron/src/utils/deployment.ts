import { Context } from 'probot';
import { MESSAGES } from '../constants/messages';
import { ENVIRONMENTS } from '../constants';
import { LATEST_STATUS_QL_QUERY } from '../constants/queries';
import { CONFIG as config } from '../constants';;
import type { LatestStatus, DeploymentStatusGroup, DeploymentGroup, Deployment, DeploymentStatus } from '../constants/types';

export const createDeployment = (
  context: Context,
  repo: string,
  owner: string,
  environment: string,
  microservice: string,
  ref: string,
  requiredContexts = [],
// eslint-disable-next-line @typescript-eslint/no-explicit-any
): Promise<any>  => {

  const user = context.payload.issue.user.login;
  context.log(MESSAGES.deploying(user, microservice, environment, ref));

  return context.github.repos.createDeployment({
    repo,
    owner,
    ref, 
    environment,
    description: 'deploy request from ci-cd',
    payload: JSON.stringify({
      microservice,
      pr: context.payload.issue.number,
      user,
    }),
    required_contexts: requiredContexts,
    transient_environment: environment !== ENVIRONMENTS.production,
    headers: {
      'Accept': 'application/vnd.github.ant-man-preview+json',
    }
  });
};




/**
 * checks to see if there are already pending deployments to this enviornment that are not the current ref
 * for shared enviornments like prod or test, you do not want to trigger a deployment workflow if one is already
 * in place
 * @param context 
 * @param ref 
 * @param env 
 * @param repo 
 * @param owner 
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isTherePendingDeploymentForEnvironment = async (context: Context, ref: string, env: string, repo: string, owner: string): Promise<DeploymentStatus[]> => {

  const data = await context.github.graphql(LATEST_STATUS_QL_QUERY,
  {
    repo,
    owner,
    maxLookup: config.maxDeploymentsToLookupForPending,
    env
  });
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  return data.repository.deployments.edges.filter((edge: LatestStatus) => (edge.node.latestStatus === 'PENDING' && edge.node.ref.name !== ref || edge.node.latestStatus === null && edge.node.ref.name !== ref));
};


// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const getLatestEnvironmentStatusesForRef = async (context: Context, ref: string, repo: string, owner:string): Promise<any> => {
  const response = await context.github.repos.listDeployments({
    repo,
    owner,
    ref
  });

  // group deployments by environment and get latest status
  const groupedDeployments = response.data.reduce((groupedDeployments: DeploymentGroup, deployment: Deployment) => {
    const { environment } = deployment;
    if(!groupedDeployments[environment]) {
      groupedDeployments[environment] = deployment;
    }
    return groupedDeployments;
  }, {});
  // latest status is always the first one, get it for each env
  const latestStatuses = Object.keys(groupedDeployments).map(async (env: string) => {
    return {env, status: await context.github.repos.listDeploymentStatuses({repo, owner, deployment_id: groupedDeployments[env].id, per_page: 1})};
  }, {});
  const resolvedStatuses = await Promise.all(latestStatuses);

  // organize statuses against grouped environments
  return  resolvedStatuses.reduce((groupedStatuses: DeploymentGroup, state) => {
    // if there is not pending deployment status ie status.data === [] sub in a pending status
    groupedStatuses[state.env] = state.status.data[0] || { state: 'pending'};
    return groupedStatuses;
  }, {});
};

/**
 * some environments require that previous environments in the train deployed succesffuly
 * for example you wouldn't want to deploy to prod unless the ref had been deployed to dev and test
 * @param requiredEnvironments 
 * @param deploymentStatuses 
 * @returns {boolean}
 */
export const isEnvironmentAllowedToDeploy = (requiredEnvironments: string[], deploymentStatuses: DeploymentStatusGroup): boolean => {
  if(!requiredEnvironments || requiredEnvironments.length === 0) return true;

  return requiredEnvironments.every(env => deploymentStatuses[env].state === 'success' );
};
