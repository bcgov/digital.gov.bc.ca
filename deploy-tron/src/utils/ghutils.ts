import { Context } from 'probot';
import { CONFIG as config } from '../constants';
import {  PrStatusMessage, DeploymentStatus, RepoOwner } from '../constants/types';

/**
 * returns the repo and owner from the context
 * @param context
 * @returns {Object}
 */
export const getRepoAndOwnerFromContext = (context: Context): RepoOwner => {
  const { repository } = context.payload;
  const {
    name: repo,
    owner: { login: owner },
  } = repository;

  return { repo, owner };
};

export const getHeadRefFromPr = async (context: Context): Promise<string> => {
  const { repo, owner } = getRepoAndOwnerFromContext(context);

  const res = await context.github.pulls.get({owner,
    repo,
    pull_number: context.payload.issue.number
  });

  return res.data.head.ref;
};
/**
 * checks if commenter can perform action
 * @param context
 * @param allowedRoles
 */
export const isCommenterAllowedToAction = async (
  context: Context,
): Promise<boolean> => {
  const { repo, owner } = getRepoAndOwnerFromContext(context);
  // get user name and see if they are a member
  const { login: username } = context.payload.comment.user;
  // get their permissions for this repo
  const { data } = await context.github.repos.getCollaboratorPermissionLevel({
    username,
    repo,
    owner,
  });
  // get allowable permissions to perform the action
  // no matter what, read users should never be allowed to trigger changes
  return (
    config.validGithubRoles.includes(data.permission) &&
    data.permission !== 'read'
  );
};

/**
 * checks if comment came froma pull request
 * @param context
 * @returns {Boolean}
 */
export const isCommentFromPr = (context: Context): boolean => {
  const { issue } = context.payload;

  return Object.prototype.hasOwnProperty.call(issue, 'pull_request');
};

/**
 * checks if comment is a bot command
 * @param context
 * @returns {Boolean}
 */
export const isBotCommand = (context: Context, botCommand: string): boolean => {
  const { body } = context.payload.comment;
  const re = new RegExp(`^${botCommand}`);

  return re.test(body.trim());
};


export const createComment = (context: Context, body: string): Promise<unknown> => {
  const params = context.issue({ body });
  return context.github.issues.createComment(params);
};


export const extractPrsThatArePendingForComment = (deployments: DeploymentStatus[]): PrStatusMessage[] => {
  // group deployment by ref
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const prs = deployments.reduce((group: any, deployment) => {
    const branch = deployment.node.ref.name;
    if(!group[branch]) {
      let payload;
      try {
        // for some reason the graphql api returns the json string with way more \ then is needed
        // VERY VERY wierd but this is the only way this works
        payload = JSON.parse(JSON.parse(JSON.parse(deployment.node.payload)));
      } catch (e) {
        payload = {
          pr: null,
        };
      }
      group[branch] = {
        ...deployment.node, 
        payload
      };
    }
    return group;
  }, {});

  return Object.keys(prs).map(branch => {
    return ({
      pr: prs[branch].payload.pr,
      branch,
      state: prs[branch].latestStatus !== null ? prs[branch].latestStatus.state : 'no status found'
    });
  });
};


