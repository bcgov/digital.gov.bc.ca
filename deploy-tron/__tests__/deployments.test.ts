/* eslint-disable @typescript-eslint/no-explicit-any */
import {  Context } from 'probot';
import { pullRequestComment } from '../__fixtures__/pull_request_comment';
import { github } from './helpers';
import { deploymentStatusesPending, deploymentStatusesPendingWithSameRef, deploymentsForRef, deploymentStatusForDeployment, deploymentStatusesSuccess } from '../__fixtures__/deployments';
import { isTherePendingDeploymentForEnvironment, getLatestEnvironmentStatusesForRef, isEnvironmentAllowedToDeploy } from '../src/utils/deployment';
import config from '../__fixtures__/config.json';


describe('Deployment Helpers', () => {


  test('isTherePendingDeploymentForEnvironment returns true when there is a pending deployent', async () => {
    const context = new Context(pullRequestComment, github as any, {} as any);

    github.graphql.mockReturnValueOnce(Promise.resolve(deploymentStatusesPending));
    const isPending = await isTherePendingDeploymentForEnvironment(context,'master', 'development', '123123', 'bar');
    expect(isPending).toMatchSnapshot();
  });

  test('isTherePendingDeploymentForEnvironment returns [] when there is a pending deployent and ref is the same', async () => {
    const context = new Context(pullRequestComment, github as any, {} as any);
    
    github.graphql.mockReturnValueOnce(Promise.resolve(deploymentStatusesPendingWithSameRef));
    const isPending = await isTherePendingDeploymentForEnvironment(context, deploymentStatusesPendingWithSameRef.repository.deployments.edges[0].node.ref.name, 'development', 'bar', 'owner');
    expect(isPending).toEqual([]);
  });

  test('isTherePendingDeploymentForEnvironment returns false when there is a non pending deployent', async () => {
    const context = new Context(pullRequestComment, github as any, {} as any);
    
    github.graphql.mockReturnValueOnce(Promise.resolve(deploymentStatusesSuccess));
    const isPending = await isTherePendingDeploymentForEnvironment(context,'master', 'development', 'foo', 'bar');
    expect(isPending).toEqual([]);
  });

  test('getLatestEnvironmentStatusesForRef returns statuses grouped by env', async () => {
    const context = new Context(pullRequestComment, github as any, {} as any);

    github.repos.listDeployments.mockReturnValue(Promise.resolve({data: deploymentsForRef}));
    github.repos.listDeploymentStatuses.mockReturnValue(Promise.resolve({data: deploymentStatusForDeployment}));

    const statuses = await getLatestEnvironmentStatusesForRef(context, 'foo', 'bar', 'baz');
    expect(statuses).toMatchSnapshot();
  });


  test('isEnvironmentAllowedToDeploy returns true when required env is empty', () => {
    expect(isEnvironmentAllowedToDeploy([], { production: { state: 'success' }})).toBe(true);
  });

  test('isEnvironmentAllowedToDeploy returns true when required env are met', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const environments = config.requiredEnvironments['prod'];
    const statuses = { 
      production: { state: 'success' },
      staging: { state: 'success' },
      development: { state: 'success' },
    };

    expect(isEnvironmentAllowedToDeploy(environments, statuses)).toBe(true);
  });
});
