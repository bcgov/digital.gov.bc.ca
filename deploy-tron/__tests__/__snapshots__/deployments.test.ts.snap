// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deployment Helpers getLatestEnvironmentStatusesForRef returns statuses grouped by env 1`] = `
Object {
  "development": Object {
    "created_at": "2020-08-07T22:24:20Z",
    "creator": Object {
      "avatar_url": "https://avatars1.githubusercontent.com/u/21046727?v=4",
      "events_url": "https://api.github.com/users/patricksimonian/events{/privacy}",
      "followers_url": "https://api.github.com/users/patricksimonian/followers",
      "following_url": "https://api.github.com/users/patricksimonian/following{/other_user}",
      "gists_url": "https://api.github.com/users/patricksimonian/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/patricksimonian",
      "id": 3434,
      "login": "patricksimonian",
      "node_id": "MDQ6VXNlcjIxMDQ2NzI3",
      "organizations_url": "https://api.github.com/users/patricksimonian/orgs",
      "received_events_url": "https://api.github.com/users/patricksimonian/received_events",
      "repos_url": "https://api.github.com/users/patricksimonian/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/patricksimonian/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/patricksimonian/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/patricksimonian",
    },
    "deployment_url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/3434",
    "description": "",
    "environment": "production",
    "id": 1312,
    "node_id": "MDE2OkRlcGxveW1lbnRTdGF0dXMzNjMzOTA1MzQ=",
    "performed_via_github_app": null,
    "repository_url": "https://api.github.com/repos/bcgov/devhub-app-web",
    "state": "success",
    "target_url": "",
    "updated_at": "2020-08-07T22:24:20Z",
    "url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/123/statuses/23434",
  },
  "staging": Object {
    "created_at": "2020-08-07T22:24:20Z",
    "creator": Object {
      "avatar_url": "https://avatars1.githubusercontent.com/u/21046727?v=4",
      "events_url": "https://api.github.com/users/patricksimonian/events{/privacy}",
      "followers_url": "https://api.github.com/users/patricksimonian/followers",
      "following_url": "https://api.github.com/users/patricksimonian/following{/other_user}",
      "gists_url": "https://api.github.com/users/patricksimonian/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/patricksimonian",
      "id": 3434,
      "login": "patricksimonian",
      "node_id": "MDQ6VXNlcjIxMDQ2NzI3",
      "organizations_url": "https://api.github.com/users/patricksimonian/orgs",
      "received_events_url": "https://api.github.com/users/patricksimonian/received_events",
      "repos_url": "https://api.github.com/users/patricksimonian/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/patricksimonian/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/patricksimonian/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/patricksimonian",
    },
    "deployment_url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/3434",
    "description": "",
    "environment": "production",
    "id": 1312,
    "node_id": "MDE2OkRlcGxveW1lbnRTdGF0dXMzNjMzOTA1MzQ=",
    "performed_via_github_app": null,
    "repository_url": "https://api.github.com/repos/bcgov/devhub-app-web",
    "state": "success",
    "target_url": "",
    "updated_at": "2020-08-07T22:24:20Z",
    "url": "https://api.github.com/repos/bcgov/devhub-app-web/deployments/123/statuses/23434",
  },
}
`;

exports[`Deployment Helpers isTherePendingDeploymentForEnvironment returns true when there is a pending deployent 1`] = `
Array [
  Object {
    "node": Object {
      "latestStatus": "PENDING",
      "ref": Object {
        "name": "foo",
      },
    },
  },
]
`;
